modules/effects/src/main/scala/dev/profunktor/redis4cats/effects.scala

```diff
@@ -40,6 +40,46 @@ object effects {
   final case class ZRange[V](start: V, end: V)
   final case class RangeLimit(offset: Long, count: Long)
 
+  sealed trait ScriptOutputType[-V, +R] {
```

Comment:
I liked the previous encoding more. I think we could make it work using the Aux pattern.

Something like:

```scala
sealed trait ScriptOutputType[V] {
  type R
  // other stuff
}

object ScriptOutputType {
  import dev.profunktor.redis4cats.JavaConversions._
 
  type Aux[V0, R0] = ScriptOutputType[V0] { type R = R0 }

  def Boolean[V]: ScriptOutputType.Aux[V, Boolean] = new ScriptOutputType[V] {
    type R = Boolean
    // more stuff
  }

}
```

The methods would no longer need the type parameter `R`:

```scala
def eval(script: String, output: ScriptOutputType[V], keys: K*): F[output.R]
```

The issue with overloading would probably still remain so I guess avoiding overloading would be best. 

We could have `eval`, `evalMany`, `evalSha` and `evalShaMany`, for instance.

Reply:
Overloading is eliminated, the methods that take values now have the `WithValues` suffix.

As for the type problem, I couldn't get it to work properly at call site. I already tried yesterday to eliminate all type params via an `Aux` (that doesn't work for sure) and now again with your proposal. It compiles all fine inside the implementation, but fails at call site as the compiler can't prove that `output.Return <: Long` (or whatever you expect). I'd love to see it work as I don't like the `[R]` param either, but don't know how to resolve it. The `Aux` pattern still helps a little, check 0f4cee1 (second to last commit).

I tried a type class encoding though and like syntactically a bit more: 5aedd5f. So instead of having pass `ScriptOutputType.XYZ` as a param, one would only need to always specify the expected return type (which is usually shorter and doesn't require reading docs):
`cmd.eval[Long]("return 42")`

 This encoding also allows to add more and more generic instances (for example, any collection instead of `List`, more numeric types). Please note it's a `sealed trait` still as I don't think exposing all the internals of `ScriptOutputType` is a good idea, so users won't be able to define their own instances. What do you think @gvolpe?

Reply:
> Overloading is eliminated, the methods that take values now have the WithValues suffix.

Awesome! :+1: 

> As for the type problem, I couldn't get it to work properly at call site. I already tried yesterday to eliminate all type params via an Aux (that doesn't work for sure) and now again with your proposal. It compiles all fine inside the implementation, but fails at call site as the compiler can't prove that output.Return <: Long (or whatever you expect). I'd love to see it work as I don't like the [R] param either, but don't know how to resolve it. The Aux pattern still helps a little, check 0f4cee1 (second to last commit).

Right, that fails because the encoding is not correct. It should be something like this:

```scala
trait Scripting[F[_], K, V] {
  def eval(script: String, output: ScriptOutputType[V], keys: K*): F[output.R]
  // other methods
}

sealed trait ScriptOutputType[V] {
  type R
  private[redis4cats] type Underlying[U]
  private[redis4cats] val outputType: JScriptOutputType
  private[redis4cats] def convert[U](in: Underlying[U]): R
}

object ScriptOutputType {
  import dev.profunktor.redis4cats.JavaConversions._

  type Aux[A0, R0] = ScriptOutputType[A0] { type R = R0 }

  def Boolean[V]: ScriptOutputType.Aux[V, Boolean] = new ScriptOutputType[V] {
    type R = Boolean
    private[redis4cats] type Underlying[_] = java.lang.Boolean
    override private[redis4cats] val outputType = JScriptOutputType.BOOLEAN
    override private[redis4cats] def convert[_](in: java.lang.Boolean): Boolean = scala.Boolean.box(in)
  }

  // other values
}
```

I just gave it a try and it seems to work fine with both Scala 2.12.x and 2.13.x.

> I tried a type class encoding though and like syntactically a bit more: 5aedd5f. So instead of having pass ScriptOutputType.XYZ as a param, one would only need to always specify the expected return type (which is usually shorter and doesn't require reading docs):
cmd.eval[Long]("return 42")

I'm :-1: on having `ScriptOutputType`'s values as implicits. I'd rather have the user provide the appropriate value and the API have the return type inferred.

> Please note it's a sealed trait

Yeah that's always good.

Reply:
Thanks a lot for your quick and helpful responses 👍 

The encoding you propose works fine in theory and I had it early on, but encountered some type inference problems (as described above) and took them for a problem of the encoding itself. After spending some time yesterday trying to make it work with the tests I wrote, I found this Scala bug: https://github.com/scala/bug/issues/11488. For us, this bug means that the type inference works only if exactly one key or value is passed as varargs, not zero or more than one.

I changed the `eval` methods to take their parameters as `List`s instead of varargs, now everything works as expected and is good to go from my side.

Reply:
Oh that's interesting! 

So if varargs were the issue I wonder if we could bring the overloaded `eval` and `evalSha` methods back? :thinking: I think it looks better. 

Thanks for being patient! :pray:

Reply:
Brought back the overloaded methods and fixed the license issue 😃 

Build fails due some weird timeout (had that sometimes before as well), locally everything runs fine.

Reply:
Thanks @ybasket ! Yeah, the tests are flaky in GH actions for some reason. I'll restart the job and merge this PR once we get the green light!